2009-03-27  Adrien DELLE CAVE <decryptus@proformatique.com>

	pf_chan_dahdi_pripreferred.patch
	$Revision$
	$Date$

	Added option pripreferred in DAHDI channel.

	Indicated if PRI B channels are preferred.
	If both priexclusive and pripreferred are TRUE, the behavior will be the same
	as if only priexclusive were TRUE.

Index: asterisk-1.4.24/channels/chan_dahdi.c
===================================================================
--- asterisk-1.4.24.orig/channels/chan_dahdi.c	2009-05-27 12:58:33.000000000 +0200
+++ asterisk-1.4.24/channels/chan_dahdi.c	2009-05-27 12:58:39.000000000 +0200
@@ -589,6 +589,13 @@
 	 */
 	unsigned int priexclusive:1;
 	/*!
+	 * \brief TRUE if PRI B channels are never exclusively selected.
+	 * \note If both priexclusive and pripreferred are TRUE, the behavior will be the same
+	 *   as if only priexclusive were TRUE.
+	 * \note Set from the "pripreferred" value read in from chan_dahdi.conf
+	 */
+	unsigned int pripreferred:1;
+	/*!
 	 * \brief TRUE if we will pulse dial.
 	 * \note Set from the "pulsedial" value read in from chan_dahdi.conf
 	 */
@@ -2467,6 +2474,8 @@
 		/* Add support for exclusive override */
 		if (p->priexclusive)
 			exclusive = 1;
+		else if (p->pripreferred)
+			exclusive = 0;
 		else {
 		/* otherwise, traditional behavior */
 			if (p->pri->nodetype == PRI_NETWORK)
@@ -7942,6 +7951,7 @@
 		tmp->use_callingpres = conf->chan.use_callingpres;
 		tmp->priindication_oob = conf->chan.priindication_oob;
 		tmp->priexclusive = conf->chan.priexclusive;
+		tmp->pripreferred = conf->chan.pripreferred;
 		if (tmp->usedistinctiveringdetection) {
 			if (!tmp->use_callerid) {
 				ast_log(LOG_NOTICE, "Distinctive Ring detect requires 'usecallerid' be on\n");
@@ -11762,6 +11772,8 @@
 						v->value, v->lineno);
 			} else if (!strcasecmp(v->name, "priexclusive")) {
 				confp->chan.priexclusive = ast_true(v->value);
+			} else if (!strcasecmp(v->name, "pripreferred")) {
+				confp->chan.pripreferred = ast_true(v->value);
 			} else if (!strcasecmp(v->name, "internationalprefix")) {
 				ast_copy_string(confp->pri.internationalprefix, v->value, sizeof(confp->pri.internationalprefix));
 			} else if (!strcasecmp(v->name, "nationalprefix")) {
